### PR TITLE
drivers: ad559x: implement double range option 

### PR DESCRIPTION
**Pull request by jeppenodgaard**
Add support for double range (2 x Vref) for both ADC and DAC.

### CODE REVIEW

**Comment by jeppenodgaard** [comment]
Rebase to restart CI which seemed stuck.

File: drivers/dac/dac_ad559x.c [hunk 1]
```diff
@@ -99,6 +113,7 @@ static int dac_ad559x_init(const struct device *dev)
 #define DAC_AD559X_DEFINE(inst)                                                                    \
 	static const struct dac_ad559x_config dac_ad559x_config##inst = {                          \
 		.mfd_dev = DEVICE_DT_GET(DT_INST_PARENT(inst)),                                    \
+		.range_double = DT_INST_PROP_OR(inst, range_double, false),                        \
```
Comment (aasinclair) [inline drivers/dac/dac_ad559x.c:116]: Boolean properties don't need to use _OR.

```suggestion
		.range_double = DT_INST_PROP(inst, range_double),                        \
```

File: drivers/adc/adc_ad559x.c [hunk 2]
```diff
@@ -272,24 +286,29 @@ static int adc_ad559x_init(const struct device *dev)
...
-	.ref_internal = AD559X_ADC_VREF_MV,
-};
+/* clang-format off */
+#define ADC_AD559X_DRIVER_API(inst)                                                                \
+	static const struct adc_driver_api adc_ad559x_api##inst = {                                \
+		.channel_setup = adc_ad559x_channel_setup,                                         \
+		.read = adc_ad559x_read,                                                           \
+		IF_ENABLED(CONFIG_ADC_ASYNC, (.read_async = adc_ad559x_read_async,))               \
+		.ref_internal = COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, range_double), (2), (1)) * \
+				AD559X_ADC_VREF_MV,                                                \
```
Comment (aasinclair) [inline drivers/adc/adc_ad559x.c:296]: You can't use HAS_PROP with a boolean, maybe try something like:

```suggestion
		.ref_internal = AD559X_ADC_VREF_MV * (1 + DT_INST_PROP(inst, range_double)),       \

```

File: drivers/adc/adc_ad559x.c [hunk 3]
```diff
@@ -272,24 +286,29 @@ static int adc_ad559x_init(const struct device *dev)
...
 
-static const struct adc_driver_api adc_ad559x_api = {
-	.channel_setup = adc_ad559x_channel_setup,
-	.read = adc_ad559x_read,
-#ifdef CONFIG_ADC_ASYNC
-	.read_async = adc_ad559x_read_async,
-#endif
-	.ref_internal = AD559X_ADC_VREF_MV,
-};
+/* clang-format off */
```
Comment (aasinclair) [inline drivers/adc/adc_ad559x.c:289]: Why the clang-format disable?
If it is because of the weird formatting after IF_ENABLED, you could try moving this section to the end of the struct.

**Review by aasinclair — DISMISSED** [review]

**Comment by jeppenodgaard** [comment]
Thanks for the review @aasinclair.
I should have addressed your comments.
The clang-format is still not perfect, but now I think your suggestion and the format is better than disabling.

File: include/zephyr/drivers/mfd/ad559x.h [hunk 4]
```diff
@@ -21,6 +21,8 @@ extern "C" {
 #define AD559X_REG_GPIO_SET       0x09U
 #define AD559X_REG_GPIO_INPUT_EN  0x0AU
 #define AD559X_REG_PD_REF_CTRL    0x0BU
+#define AD559X_REG_GPIO_OD_CONFIG 0x0CU
```
Comment (bbilas) [inline include/zephyr/drivers/mfd/ad559x.h:25]: This is not related to this commit at all.

File: drivers/dac/dac_ad559x.c [hunk 5]
```diff
@@ -88,6 +89,19 @@ static int dac_ad559x_init(const struct device *dev)
...
 	}
 
+	if (config->range_double) {
+		uint16_t tmp;
+
+		ret = mfd_ad559x_read_reg(config->mfd_dev, AD559X_REG_GEN_CTRL, 0, &tmp);
+		if (ret < 0) {
+			return ret;
+		}
+		ret = mfd_ad559x_write_reg(config->mfd_dev, AD559X_REG_GEN_CTRL, tmp);
```
Comment (bbilas) [inline drivers/dac/dac_ad559x.c:99]: How is it supposed to work? You do a read and write the same register without any modification.

File: drivers/dac/dac_ad559x.c [hunk 6]
```diff
@@ -88,6 +89,19 @@ static int dac_ad559x_init(const struct device *dev)
 		return -ENODEV;
 	}
 
+	if (config->range_double) {
+		uint16_t tmp;
+
+		ret = mfd_ad559x_read_reg(config->mfd_dev, AD559X_REG_GEN_CTRL, 0, &tmp);
+		if (ret < 0) {
+			return ret;
+		}
```
Comment (bbilas) [inline drivers/dac/dac_ad559x.c:98]: Add a `nl`.

**Comment by jeppenodgaard** [comment]
Thanks for the review @bbilas and especially catching the missing register update.
I also changed it so the bit is always modified based on the dts property.
Otherwise the bit would not be cleared when `range-double` is not present in the dts.

File: drivers/adc/adc_ad559x.c (hunk 2 above)
Comment (bbilas) [reply drivers/adc/adc_ad559x.c]: or 

<pre>
#if DT_INST_PROP(inst, range_double)
#define AD559X_REF_INTERNAL
	.ref_internal = AD559X_ADC_VREF_MV * 2,
#else
#define AD559X_REF_INTERNAL
	.ref_internal = AD559X_ADC_VREF_MV,
#endif
</pre>

**Review by bbilas — COMMENTED** [review]
You have a typo in commit message

`s/referance/reference`

**Comment by jeppenodgaard** [comment]
@bbilas I addressed the typo and your ideas, expect for the code that I already changed via @aasinclair's comment.

**Comment by bbilas** [comment]
> @bbilas I addressed the typo and your ideas, expect for the code that I already changed via @aasinclair's comment.

That's okay. Please rename the `range-double` property with `double-input-range` (which IMO. is more understandable) as the last thing and we can approve this MR.

**Comment by jeppenodgaard** [comment]
> That's okay. Please rename the `range-double` property with `double-input-range` (which IMO. is more understandable) as the last thing and we can approve this MR.

Changed to `double-input-range` for ADC and `double-output-range` for DAC.

File: drivers/adc/adc_ad559x.c [hunk 7]
```diff
@@ -272,24 +290,31 @@ static int adc_ad559x_init(const struct device *dev)
...
-static const struct adc_driver_api adc_ad559x_api = {
-	.channel_setup = adc_ad559x_channel_setup,
-	.read = adc_ad559x_read,
 #ifdef CONFIG_ADC_ASYNC
-	.read_async = adc_ad559x_read_async,
+#define ADC_AD559X_ASYNC() .read_async = adc_ad559x_read_async,
+#else
+#define ADC_AD559X_ASYNC()
 #endif
-	.ref_internal = AD559X_ADC_VREF_MV,
```
Comment (bbilas) [inline drivers/adc/adc_ad559x.c:281]: I took a look at this part again and I'm worried why you didn't do it this way:

<pre>
static const struct adc_driver_api adc_ad559x_api = {
	.channel_setup = adc_ad559x_channel_setup,
	.read = adc_ad559x_read,
#ifdef CONFIG_ADC_ASYNC
	.read_async = adc_ad559x_read_async,
#endif
	.ref_internal = AD559X_ADC_VREF_MV * (1 + DT_INST_PROP(inst, double_input_range)),
};
</pre>

Then you could remove entire `ADC_AD559X_DRIVER_API` and etc...

Comment (jeppenodgaard) [reply drivers/adc/adc_ad559x.c:281]: Because different instances can have different `.ref_internal`.

Comment (bbilas) [reply drivers/adc/adc_ad559x.c:281]: You are right, I have forgotten about that.

**Review by bbilas — APPROVED** [review]

File: drivers/mfd/mfd_ad559x_i2c.c [hunk 8]
```diff
@@ -35,7 +37,9 @@ static int mfd_ad559x_i2c_read_reg(const struct device *dev, uint8_t reg, uint8_
 
 	ARG_UNUSED(reg_data);
 
-	__ASSERT((reg & 0xf0) == 0, "reg bits [7:4] should be 0: 0x%x", reg);
+	if (reg >= AD559X_REG_SEQ_ADC || reg <= AD559X_REG_IO_TS_CONFIG) {
+		reg |= AD559X_REG_RD_POINTER;
+	}
```
Comment (ttmut) [inline drivers/mfd/mfd_ad559x_i2c.c:42]: Just being curious, could you explain what this block does? I could not figure it out from the datasheet.

Comment (ttmut) [reply drivers/mfd/mfd_ad559x_i2c.c:42]: I was actually looking at AD5592R and that was what confused me. Thanks for the explanation.

**Comment by jeppenodgaard** [comment]
> Thanks for the review @aasinclair. I should have addressed your comments. The clang-format is still not perfect, but now I think your suggestion and the format is better than disabling.

Ping @aasinclair

**Review by MaureenHelm — APPROVED** [review]
